pkg/chunk/aws/sse_config.go

```diff
@@ -0,0 +1,85 @@
+package aws
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"flag"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	// SSEKMS config type constant to configure S3 server side encryption using KMS
+	// https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingKMSEncryption.html
+	SSEKMS     = "SSE-KMS"
+	sseKMSType = "aws:kms"
+	// SSES3 config type constant to configure S3 server side encryption with AES-256
+	// https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingServerSideEncryption.html
+	SSES3     = "SSE-S3"
+	sseS3Type = "AES256"
+)
+
+// SSEParsedConfig configures server side encryption (SSE)
+// struct used internally to configure AWS S3
+type SSEParsedConfig struct {
+	ServerSideEncryption string
+	KMSKeyID             *string
+	KMSEncryptionContext *string
+}
+
+// SSEConfig configures S3 server side encryption
+// struct that is going to receive user input (through config file or CLI)
+type SSEConfig struct {
+	Type                 string `yaml:"type"`
+	KMSKeyID             string `yaml:"kms_key_id"`
+	KMSEncryptionContext string `yaml:"kms_encryption_context"`
+}
+
+// RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet
+func (cfg *SSEConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.StringVar(&cfg.Type, prefix+"type", "", "Enable AWS Server Side Encryption. Only SSE-S3 and SSE-KMS are supported")
+	f.StringVar(&cfg.KMSKeyID, prefix+"kms-key-id", "", "KMS Key ID used to encrypt objects in S3")
+	f.StringVar(&cfg.KMSEncryptionContext, prefix+"kms-encryption-context", "", "KMS Encryption Context used for object encryption. It expects a JSON as a string.")
+}
+
+// NewSSEParsedConfig creates a struct to configure server side encryption (SSE)
+func NewSSEParsedConfig(sseType string, kmsKeyID string, kmsEncryptionContext string) (*SSEParsedConfig, error) {
+	switch sseType {
+	case SSES3:
+		return &SSEParsedConfig{
+			ServerSideEncryption: sseS3Type,
+		}, nil
+	case SSEKMS:
+		if kmsKeyID == "" {
+			return nil, errors.New("KMS key id must be passed when SSE-KMS encryption is selected")
+		}
+
+		parsedKMSEncryptionContext, err := parseKMSEncryptionContext(kmsEncryptionContext)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to parse KMS encryption context")
+		}
+
+		return &SSEParsedConfig{
+			ServerSideEncryption: sseKMSType,
+			KMSKeyID:             &kmsKeyID,
+			KMSEncryptionContext: parsedKMSEncryptionContext,
+		}, nil
+	default:
+		return nil, errors.New("SSE type is empty or invalid")
+	}
+}
+
+func parseKMSEncryptionContext(kmsEncryptionContext string) (*string, error) {
+	if kmsEncryptionContext == "" {
+		return nil, nil
+	}
+
+	jsonKMSEncryptionContext, err := json.Marshal(json.RawMessage(kmsEncryptionContext))
```

Comment:
If `kmsEncryptionContext` is already a JSON as a string, why do we need to re-marshal it? Can't we just use it as is and base64 encode it?

Reply:
Using `json.Marshal`, I can validate if the string is a valid `JSON`. Giving the user a better error message in case they pass an invalid `JSON`.

Reply:
Ok, fair enough. Please add a code comment to explain it.

Reply:
Can we also add a unit test with invalid JSON?